### ISSUE TITLE
Loding additional CSS by JS in IE when caching is on

### ISSUE THREAD

**Issue by ymer94** [pojome/elementor#5617]
## Prerequisites

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [x] The issue still exists against the latest stable version of Elementor.

## Description

When I opened the site in IE (after updating Elementor a few days ago) there were som issues with the styling of some elements (some would show up like normal, some with weird styling, like margins/paddings, and some would not show up at all, due to an exception i could find in the browsers console.

After clearing all caches and disabling them the site worked fine even in IE and after some further investigation I found out that _/elementor/assets/js/frontend.js_ tries to load a specific CSS file designed for IE, _/elementor/assets/css/frontend-msie.min.css_, but it tries to load it directly after _/elementor/assets/css/frontend.min.css_ using that link tags ID attribute. Since that link tag is long gone, along with it's ID, that caused the exception. I have temporarily fixed this by excluding that CSS file to be processed by the plugin that handles the CSS optimization (in my case Fast Velocity Minify).

## Suggested solution

The below code is what I found that targets the _elementor-frontend-css_ ID and because of that breaks the page in IE unless I exclude the _/elementor/assets/css/frontend.min.css_ from being minified.

```javascript
var addIeCompatibility = function addIeCompatibility() {
	var isIE = 'Microsoft Internet Explorer' === navigator.appName || !!navigator.userAgent.match(/Trident/g) || !!navigator.userAgent.match(/MSIE/g) || !!navigator.userAgent.match(/rv:11/);

	if (!isIE) {
		return;
	}
	elements.$body.addClass('elementor-msie');

	var $frontendCss = jQuery('#elementor-frontend-css'),
	msieCss = $frontendCss[0].outerHTML.replace('css/frontend', 'css/frontend-msie').replace('elementor-frontend-css', 'elementor-frontend-msie-css');

	$frontendCss.after(msieCss);
};
```

To solve this problem I have three suggestions (I don't know if all of them can be done but as I said it's some suggestions):

1. Move that functionality to the PHP code and enqueue the CSS file like all others (using $_SERVER['HTTP_USER_AGENT'] or some other method of detecting IE on the server side).
2. Insert a some other tag in the head (from the PHP) to target and insert the link tags after that tag.
3. Just append the link tags to the head tag (maybe not the best solution but it won't break the page as much as an ID that can't be found and it would be easier to debug if it does break some design).

## Environment

Don't know if you need this since I already have debugged it and made a temporary solution but I'll include it anyways in case you need it.

<details>
<summary>System Info</summary>

== Server Environment ==
Operating System: Linux
Software: Apache
MySQL version: 10.1.31
PHP Version: 7.2.8
PHP Max Input Vars: 1000
PHP Max Post Size: 128M
GD Installed: Yes
ZIP Installed: Yes
Write Permissions: All right
Elementor Library: Connected

== WordPress Environment ==
Version: 4.9.8
Site URL: https://fagerstamotorforum.se
Home URL: https://fagerstamotorforum.se
WP Multisite: No
Max Upload Size: 16 MB
Memory limit: 40M
Permalink Structure: /%postname%
Language: en-US
Timezone: Europe/Stockholm
Debug Mode: Inactive

== Theme ==
Name: Make Child Fagerstamotorforum
Version: 1.9.13
Author: Ymer Nordström
Child Theme: Yes
Parent Theme Name: Make
Parent Theme Version: 1.9.13
Parent Theme Author: The Theme Foundry

== User ==
Role: administrator
WP Profile lang: sv_SE
User Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.106 Safari/537.36 OPR/55.0.2994.44

== Active Plugins ==
All In One SEO Pack
Version: 2.7.3
Author: Michael Torbert

Attachment Pages Redirect
Version: 1.1
Author: Samuel Aguilera

Breadcrumb NavXT
Version: 6.1.0
Author: John Havlik

Duplicator Pro
Version: 3.7.3.1
Author: Snap Creek

Elementor
Version: 2.2.0
Author: Elementor.com

Fast Velocity Minify
Version: 2.3.5
Author: Raul Peixoto

jQuery Updater
Version: 3.3.1
Author: Ramoonus

Optimize Database after Deleting Revisions
Version: 4.6.3
Author: CAGE Web Design | Rolf van Gelder, Eindhoven, The Netherlands

Permalink Manager Lite
Version: 2.0.6.3
Author: Maciej Bis

reSmush.it Image Optimizer
Version: 0.1.16
Author: reSmush.it

Safe SVG
Version: 1.6.1
Author: Daryll Doyle

Sticky Menu (or Anything!) on Scroll
Version: 2.1.1
Author: Mark Senff

W3 Total Cache
Version: 0.9.7
Author: Frederick Townes

Wordfence Security
Version: 7.1.11
Author: Wordfence

WP Mail, Email Logs, Gmail SMTP, PHP Mailer - Mail Bank
Version: 3.0.64
Author: Tech Banker

</details>

**Comment by bainternet** [comment]
@ymer94 

Thanks for reporting,

Not really sure what the issue is, but a few notes:

> Move that functionality to the PHP code and enqueue the CSS file like all others (using $_SERVER['HTTP_USER_AGENT'] or some other method of detecting IE on the server side).

This can't be done is the site uses any type of server-side cache.

> Insert a some other tag in the head (from the PHP) to target and insert the link tags after that tag.

Not always possible since you need to render the page to know which css file to include, something that is not always known in advance.

also why would you minify frontend.min.css when its already minified?

anyway, this is not something that needs fixing.

thanks again.

**Comment by ymer94** [comment]
Well it's not specifically that file I wan't to minimize but all the CSS in total of the site. Right now I've set Fast Velocity Minify to take all CSS files that a page needs and merge them into as few files as possible and then inline those files in the html code. This is to speed up page loading times but right now I have to exclude frontend.min.css for it to work in IE.

I know for a fact that moving it to the PHP code would solve the problem (at least for the plugin I'm using) since it optimizes the page upon request of that page. This means that if some other source than a browser would make a request for a page the frontend-msie.min.css would not be included. However if an IE browser would load that page after it's been optimized Fast Velocity Minify would detect that a new CSS file (that it hasen't optimized yet) is enqueued and optimize that one too for IE. The only thing about this method I'm uncertain of is if $_SERVER['HTTP_USER_AGENT'] is the best way to detect en IE browser since if can be spoofed.

What I meant by another tag in the head is i.e. a meta tag that later could be the target for the jQuery to insert the CSS link tags after. That meta tag could be there regardless of browser and only used as a target when the JS detects an IE browser.

Hope I clarified a bit of what I meant and thanks for the quick reply, but I disagree with it not being something that needs fixing since it has a potential to break some sites when viewed in IE depending on how they implement caching and page speed optimization.

There is also other options like targeting IE browsers in the CSS file specifically or using a specific ID/Class or other attribute in the CSS rules and only add that with attribute by JS when it's being rendered in IE.

**Comment by mcrowell** [comment]
I'm running into the same issue when using autoptimize & cache enabler. Internet Explorer doesn't render the inline CSS correctly. 

Microsoft Edge renders correctly.

**Comment by ymer94** [comment]
@mcrowell As I said my solution was, try  manually excluding the /elementor/assets/css/frontend.min.css from all optimization plugins you have installed. It's only IE that uses the function I referenced in my original issue since Edge uses a different user agent (UA) string then it's predecessors and those stings are not checked in the if statement deciding whether to run the jQuery call or not. Not sure if I explained it good enough but if you have questions about it feel free to ask.
